File: packages/server/src/orderbook.ts
import type { Order } from '@orderbook/shared';

/**
 * OrderBook — Data-only class (no matching logic).
 *
 * Storage:
 *   • `bidsByPrice` / `asksByPrice` — Map<price, Order[]>
 *     Each bucket holds orders at that price in FIFO arrival order.
 *     Used for O(1) price-level access and as the source of truth for
 *     getBestBid / getBestAsk.
 *
 *   • `bidPriceLevels` / `askPriceLevels` — sorted number[]
 *     Compact arrays of *distinct prices only* (not individual orders).
 *     Kept sorted (bids descending, asks ascending) via binary search insert.
 *     Length = number of distinct active price levels (typically 20–50),
 *     not total order count.
 *
 *   • `ordersById` — Map<id, Order>
 *     Used for O(1) cancel and ID-based lookup.
 *
 * Complexity (p = distinct price levels, k = orders at one price level):
 *   addOrder    → O(log p + p)  binary search find + splice shift in price levels
 *   removeOrder → O(log p + p)  binary search find + splice shift in price levels
 *                 + O(k) scan within the price-level bucket (k is typically small)
 *   getBestBid  → O(1)          index 0 of bidPriceLevels → bucket[0]
 *   getBestAsk  → O(1)          index 0 of askPriceLevels → bucket[0]
 *   getTop10    → O(10)         walk first few price levels until 10 orders collected
 *
 * Previous implementation maintained full sorted Order[] arrays and called
 * .sort() on every insert (O(n log n)) and findIndex + splice on every
 * removal (O(n)), where n = total orders in the book. This caused throughput
 * to degrade visibly at high order counts. The fix separates the display path
 * (getTop10) from the matching hot path (getBestBid / getBestAsk) — only the
 * latter needs to be O(1) and it now is, without touching the full order arrays.
 */
export class OrderBook {
  // ── price-level maps (source of truth) ───────────────
  private bidsByPrice = new Map<number, Order[]>();
  private asksByPrice = new Map<number, Order[]>();

  // ── sorted price-level arrays (distinct prices only) ─
  // bids: descending (highest bid first)
  // asks: ascending  (lowest ask first)
  private bidPriceLevels: number[] = [];
  private askPriceLevels: number[] = [];

  // ── ID map ────────────────────────────────────────────
  private ordersById = new Map<string, Order>();

  // ── order count ───────────────────────────────────────
  private _size = 0;

  // ─── Add ─────────────────────────────────────────────
  addOrder(order: Order): void {
    this.ordersById.set(order.id, order);

    const priceMap = order.side === 'buy' ? this.bidsByPrice : this.asksByPrice;
    const existing = priceMap.get(order.price);

    if (existing) {
      // Price level already exists — append (preserves FIFO within level).
      existing.push(order);
    } else {
      // New price level — create bucket and record the price in sorted array.
      priceMap.set(order.price, [order]);
      if (order.side === 'buy') {
        this.insertSorted(this.bidPriceLevels, order.price, 'desc');
      } else {
        this.insertSorted(this.askPriceLevels, order.price, 'asc');
      }
    }

    this._size++;
  }

  // ─── Remove ──────────────────────────────────────────
  removeOrder(orderId: string): Order | null {
    const order = this.ordersById.get(orderId);
    if (!order) return null;

    this.ordersById.delete(orderId);

    const priceMap = order.side === 'buy' ? this.bidsByPrice : this.asksByPrice;
    const bucket = priceMap.get(order.price)!;

    // Remove by reference — avoids a second string comparison loop.
    const idx = bucket.indexOf(order);
    if (idx !== -1) bucket.splice(idx, 1);

    if (bucket.length === 0) {
      // Price level is now empty — evict from map and price levels array.
      priceMap.delete(order.price);
      if (order.side === 'buy') {
        this.removeSorted(this.bidPriceLevels, order.price, 'desc');
      } else {
        this.removeSorted(this.askPriceLevels, order.price, 'asc');
      }
    }

    this._size--;
    return order;
  }

  // ─── Queries ─────────────────────────────────────────
  getOrdersAtPrice(side: 'buy' | 'sell', price: number): Order[] {
    const priceMap = side === 'buy' ? this.bidsByPrice : this.asksByPrice;
    return priceMap.get(price) ?? [];
  }

  getBestBid(): Order | null {
    if (this.bidPriceLevels.length === 0) return null;
    // bidPriceLevels[0] is always the highest active bid price.
    return this.bidsByPrice.get(this.bidPriceLevels[0])![0] ?? null;
  }

  getBestAsk(): Order | null {
    if (this.askPriceLevels.length === 0) return null;
    // askPriceLevels[0] is always the lowest active ask price.
    return this.asksByPrice.get(this.askPriceLevels[0])![0] ?? null;
  }

  getTop10Bids(): Order[] {
    return this.getTopN(this.bidPriceLevels, this.bidsByPrice, 10);
  }

  getTop10Asks(): Order[] {
    return this.getTopN(this.askPriceLevels, this.asksByPrice, 10);
  }

  getOrderById(orderId: string): Order | null {
    return this.ordersById.get(orderId) ?? null;
  }

  get size(): number {
    return this._size;
  }

  // ─── Private helpers ─────────────────────────────────

  /**
   * Collect up to `n` orders by iterating price levels in priority order.
   * Naturally respects price-time priority: levels are pre-sorted, and
   * orders within each level are stored in FIFO arrival order.
   */
  private getTopN(
    priceLevels: number[],
    priceMap: Map<number, Order[]>,
    n: number,
  ): Order[] {
    const result: Order[] = [];
    for (const price of priceLevels) {
      if (result.length >= n) break;
      for (const order of priceMap.get(price)!) {
        result.push(order);
        if (result.length >= n) break;
      }
    }
    return result;
  }

  /**
   * Binary search insert — O(log p) to find position, O(p) for the splice shift.
   * Keeps `arr` sorted without a full re-sort on every call.
   * dir 'desc': highest first (bids).
   * dir 'asc':  lowest first  (asks).
   */
  private insertSorted(arr: number[], price: number, dir: 'asc' | 'desc'): void {
    let lo = 0,
      hi = arr.length;
    while (lo < hi) {
      const mid = (lo + hi) >> 1;
      const goRight = dir === 'desc' ? arr[mid] > price : arr[mid] < price;
      if (goRight) lo = mid + 1;
      else hi = mid;
    }
    arr.splice(lo, 0, price);
  }

  /**
   * Binary search remove — O(log p) to find, O(p) for the splice shift.
   * Only called when a price-level bucket becomes empty.
   */
  private removeSorted(arr: number[], price: number, dir: 'asc' | 'desc'): void {
    let lo = 0,
      hi = arr.length;
    while (lo < hi) {
      const mid = (lo + hi) >> 1;
      const goRight = dir === 'desc' ? arr[mid] > price : arr[mid] < price;
      if (goRight) lo = mid + 1;
      else hi = mid;
    }
    if (arr[lo] === price) arr.splice(lo, 1);
  }
}

File: packages/matching-engine-go/internal/orderbook.go
package internal

// OrderBook — Optimized Implementation (Linked Queue Per Price Level)
//
// Thread Safety:
//   This OrderBook is NOT thread-safe on its own. Thread safety is provided
//   by the semaphore in main.go which serializes all HTTP handler access.
//   This design reduces lock overhead from 10+ lock/unlock cycles per request
//   to just 1 acquire/release at the handler level.
//
// Design:
//   - map[float64]*PriceLevel where PriceLevel = doubly-linked list of orders
//   - Sorted price-level arrays (distinct prices only)
//   - No array splice overhead on removals — O(1) unlink
//
// Complexity (p = distinct price levels, k = orders at one price):
//   AddOrder    → O(log p)  binary search + O(1) append to tail
//   RemoveOrder → O(1)      direct unlink (no array shift)
//   GetBestBid  → O(1)      index 0 of bidPriceLevels → level.head
//   GetBestAsk  → O(1)      index 0 of askPriceLevels → level.head
//   GetTop10    → O(10)     walk linked lists

// ── Linked list node ─────────────────────────────────
type OrderNode struct {
	Order *Order
	Next  *OrderNode
	Prev  *OrderNode
}

// ── Price level (doubly-linked list) ─────────────────
type PriceLevel struct {
	Head  *OrderNode
	Tail  *OrderNode
	Count int
}

// NewPriceLevel creates a new empty price level
func NewPriceLevel() *PriceLevel {
	return &PriceLevel{
		Head:  nil,
		Tail:  nil,
		Count: 0,
	}
}

// Append adds an order to the tail (FIFO)
func (pl *PriceLevel) Append(order *Order) *OrderNode {
	node := &OrderNode{Order: order, Next: nil, Prev: nil}

	if pl.Tail == nil {
		// Empty list
		pl.Head = node
		pl.Tail = node
	} else {
		// Append to tail
		pl.Tail.Next = node
		node.Prev = pl.Tail
		pl.Tail = node
	}

	pl.Count++
	return node
}

// Remove removes a node from the list (O(1))
func (pl *PriceLevel) Remove(node *OrderNode) {
	if node.Prev != nil {
		node.Prev.Next = node.Next
	} else {
		// Removing head
		pl.Head = node.Next
	}

	if node.Next != nil {
		node.Next.Prev = node.Prev
	} else {
		// Removing tail
		pl.Tail = node.Prev
	}

	pl.Count--
}

// CollectOrders collects up to n orders from this level
func (pl *PriceLevel) CollectOrders(n int) []*Order {
	result := make([]*Order, 0, n)
	current := pl.Head
	for current != nil && len(result) < n {
		result = append(result, current.Order)
		current = current.Next
	}
	return result
}

// GetAllOrders returns all orders (for testing/debugging)
func (pl *PriceLevel) GetAllOrders() []*Order {
	result := make([]*Order, 0, pl.Count)
	current := pl.Head
	for current != nil {
		result = append(result, current.Order)
		current = current.Next
	}
	return result
}

// ── Main OrderBook ───────────────────────────────────
type OrderBook struct {
	// Price-level maps (linked lists per price)
	bidsByPrice map[float64]*PriceLevel
	asksByPrice map[float64]*PriceLevel

	// Sorted price-level arrays (distinct prices only)
	// bids: descending (highest bid first)
	// asks: ascending (lowest ask first)
	bidPriceLevels []float64
	askPriceLevels []float64

	// ID → node lookup for O(1) cancel
	nodesById map[string]*OrderNode

	// Order count
	size int
}

// NewOrderBook creates a new empty order book
func NewOrderBook() *OrderBook {
	return &OrderBook{
		bidsByPrice:    make(map[float64]*PriceLevel),
		asksByPrice:    make(map[float64]*PriceLevel),
		bidPriceLevels: make([]float64, 0),
		askPriceLevels: make([]float64, 0),
		nodesById:      make(map[string]*OrderNode),
		size:           0,
	}
}

// AddOrder adds an order to the book
func (ob *OrderBook) AddOrder(order *Order) {
	var priceMap map[float64]*PriceLevel
	var priceLevels *[]float64
	var ascending bool

	if order.Side == SideBuy {
		priceMap = ob.bidsByPrice
		priceLevels = &ob.bidPriceLevels
		ascending = false // bids descending
	} else {
		priceMap = ob.asksByPrice
		priceLevels = &ob.askPriceLevels
		ascending = true // asks ascending
	}

	level, exists := priceMap[order.Price]
	if !exists {
		// New price level — create and insert price into sorted array
		level = NewPriceLevel()
		priceMap[order.Price] = level
		insertSorted(priceLevels, order.Price, ascending)
	}

	// Append order to level's tail (FIFO)
	node := level.Append(order)
	ob.nodesById[order.ID] = node

	ob.size++
}

// RemoveOrder removes an order from the book by ID
func (ob *OrderBook) RemoveOrder(orderID string) *Order {
	node, exists := ob.nodesById[orderID]
	if !exists {
		return nil
	}

	order := node.Order
	delete(ob.nodesById, orderID)

	var priceMap map[float64]*PriceLevel
	var priceLevels *[]float64
	var ascending bool

	if order.Side == SideBuy {
		priceMap = ob.bidsByPrice
		priceLevels = &ob.bidPriceLevels
		ascending = false
	} else {
		priceMap = ob.asksByPrice
		priceLevels = &ob.askPriceLevels
		ascending = true
	}

	level := priceMap[order.Price]

	// Unlink from list (O(1))
	level.Remove(node)

	if level.Count == 0 {
		// Price level is now empty — evict
		delete(priceMap, order.Price)
		removeSorted(priceLevels, order.Price, ascending)
	}

	ob.size--
	return order
}

// GetBestBid returns the highest bid order (or nil if none)
func (ob *OrderBook) GetBestBid() *Order {
	if len(ob.bidPriceLevels) == 0 {
		return nil
	}

	level := ob.bidsByPrice[ob.bidPriceLevels[0]]
	if level == nil || level.Head == nil {
		return nil
	}
	return level.Head.Order
}

// GetBestAsk returns the lowest ask order (or nil if none)
func (ob *OrderBook) GetBestAsk() *Order {
	if len(ob.askPriceLevels) == 0 {
		return nil
	}

	level := ob.asksByPrice[ob.askPriceLevels[0]]
	if level == nil || level.Head == nil {
		return nil
	}
	return level.Head.Order
}

// GetTop10Bids returns the top 10 bid orders
func (ob *OrderBook) GetTop10Bids() []*Order {
	return ob.getTopN(ob.bidPriceLevels, ob.bidsByPrice, 10)
}

// GetTop10Asks returns the top 10 ask orders
func (ob *OrderBook) GetTop10Asks() []*Order {
	return ob.getTopN(ob.askPriceLevels, ob.asksByPrice, 10)
}

// GetOrderByID retrieves an order by its ID
func (ob *OrderBook) GetOrderByID(orderID string) *Order {
	node := ob.nodesById[orderID]
	if node == nil {
		return nil
	}
	return node.Order
}

// Size returns the total number of orders in the book
func (ob *OrderBook) Size() int {
	return ob.size
}

// ─── Private helpers ─────────────────────────────────

// getTopN collects up to n orders by iterating price levels in priority order.
func (ob *OrderBook) getTopN(priceLevels []float64, priceMap map[float64]*PriceLevel, n int) []*Order {
	result := make([]*Order, 0, n)

	for _, price := range priceLevels {
		if len(result) >= n {
			break
		}
		level := priceMap[price]
		remaining := n - len(result)
		result = append(result, level.CollectOrders(remaining)...)
	}

	return result
}

// insertSorted inserts a price into a sorted slice using binary search
// O(log p) to find position, O(p) for the copy shift
// ascending true: lowest first (asks)
// ascending false: highest first (bids)
func insertSorted(arr *[]float64, price float64, ascending bool) {
	slice := *arr
	lo, hi := 0, len(slice)

	for lo < hi {
		mid := (lo + hi) / 2
		var goRight bool
		if ascending {
			goRight = slice[mid] < price
		} else {
			goRight = slice[mid] > price
		}

		if goRight {
			lo = mid + 1
		} else {
			hi = mid
		}
	}

	// Insert at position lo
	slice = append(slice, 0)
	copy(slice[lo+1:], slice[lo:])
	slice[lo] = price
	*arr = slice
}

// removeSorted removes a price from a sorted slice using binary search
// O(log p) to find, O(p) for the copy shift
func removeSorted(arr *[]float64, price float64, ascending bool) {
	slice := *arr
	lo, hi := 0, len(slice)

	for lo < hi {
		mid := (lo + hi) / 2
		var goRight bool
		if ascending {
			goRight = slice[mid] < price
		} else {
			goRight = slice[mid] > price
		}

		if goRight {
			lo = mid + 1
		} else {
			hi = mid
		}
	}

	if lo < len(slice) && slice[lo] == price {
		slice = append(slice[:lo], slice[lo+1:]...)
		*arr = slice
	}
}
